src/sage/matroids/flats_matroid.pyx

```diff
@@ -0,0 +1,534 @@
+        return SS._isomorphism(OS) is not None
+
+    # representation
+
+    def _repr_(self):
+        """
+        Return a string representation of the matroid.
+
+        EXAMPLES::
+
+            sage: from sage.matroids.flats_matroid import FlatsMatroid
+            sage: M = FlatsMatroid(matroids.Uniform(6, 6)); M
+            Matroid of rank 6 on 6 elements with 64 flats
+        """
+        flats_num = sum(1 for i in self._F for F in self._F[i])
+        return Matroid._repr_(self) + " with " + str(flats_num) + " flats"
+
+    # comparison
+
+    def __hash__(self):
+        r"""
+        Return an invariant of the matroid.
+
+        This function is called when matroids are added to a set. It is very
+        desirable to override it so it can distinguish matroids on the same
+        groundset, which is a very typical use case!
+
+        .. WARNING::
+
+            This method is linked to __richcmp__ (in Cython) and __cmp__ or
+            __eq__/__ne__ (in Python). If you override one, you should
+            (and in Cython: MUST) override the other!
+
+        EXAMPLES::
+
+            sage: from sage.matroids.flats_matroid import FlatsMatroid
+            sage: M = FlatsMatroid(matroids.catalog.Vamos())
+            sage: N = FlatsMatroid(matroids.catalog.Vamos())
+            sage: hash(M) == hash(N)
+            True
+            sage: O = FlatsMatroid(matroids.catalog.NonVamos())
+            sage: hash(M) == hash(O)
+            False
+        """
+        flats = frozenset([F for i in self._F for F in self._F[i]])
+        return hash(tuple([self._groundset, flats]))
+
+    def __richcmp__(left, right, int op):
+        r"""
+        Compare two matroids.
+
+        We take a very restricted view on equality: the objects need to be of
+        the exact same type (so no subclassing) and the internal data need to
+        be the same. For FlatsMatroids, this means that the groundsets and the
+        dictionaries of flats of the two matroids are equal.
+
+        EXAMPLES::
+
+            sage: from sage.matroids.flats_matroid import FlatsMatroid
+            sage: M = FlatsMatroid(matroids.catalog.Pappus())
+            sage: N = FlatsMatroid(matroids.catalog.NonPappus())
+            sage: M == N
+            False
+            sage: N = Matroid(M.bases())
+            sage: M == N
+            False
+        """
+        cdef FlatsMatroid lt, rt
+        if op not in [Py_EQ, Py_NE]:
+            return NotImplemented
+        if type(left) is not type(right):
+            return NotImplemented
+        lt = <FlatsMatroid> left
+        rt = <FlatsMatroid> right
+        if lt.groundset() != rt.groundset():
+            return rich_to_bool(op, 1)
+        if lt.full_rank() != rt.full_rank():
+            return rich_to_bool(op, 1)
+        return richcmp(lt._F, rt._F, op)
+
+    # copying, loading, saving
+
+    def __copy__(self):
+        """
+        Create a shallow copy.
```

Comment:
I know that you are keeping this parallel to the existing classes, but for immutable objects, both `__copy__` and `__deepcopy__` really should just return `self`. 
See #13811

Reply:
Done. I think I will have to open a separate PR to do this for other classes too. In fact, it would suffice to only do it in the main abstract matroid class. This will save many (redundant) lines of code.

Reply:
Yes, that's a good idea.